Merge #27, #29, #26 & various new features

charlesdaniels opened the pull request:
New features include

* Lots of automated testing scripts
* Automation for uploading to PyPi
* GitHub release uploading for TravisCI (unclear how well this works just yet)
* `setup.py` works correctly and can be used to install under Python 2 or Python 3
* `--encode` or `--decode` can now be inferred based on context in some cases

The only new features which do not have test cases are

* `--editor` implies `--decode`
* editor detection
* reading data to decode from an editor

I am comfortable with the level of quality and validation in this changeset if @jyn514  is.

jyn514 committed:
Add Github release uploading

Closes https://github.com/charlesdaniels/bitshuffle/issues/11.
Needs testing.

jyn514 committed:
Add Github release uploading

Closes https://github.com/charlesdaniels/bitshuffle/issues/11.
Needs testing.

jyn514 committed:
Add pypi metadata

Add version to program
Add Charles as author
Add github page as homepage
Add MANIFEST, setup.cfg

jyn514 committed:
Add pypi metadata

Add version to program
Add Charles as author
Add github page as homepage
Add MANIFEST, setup.cfg

jyn514 committed:
Add build script

Fix typo in setup.py

jyn514 committed:
Add build script

Fix typo in setup.py

bx-r0 committed:
Changes;
-Added --version flag

bx-r0 committed:
Changes;
-Added the inference for the Encode and Decode flags.
-Added a check for the existence of a file

bx-r0 committed:
-Small changes to make Travis happy

bx-r0 committed:
Fix for W503 - Line break before binary operator

jyn514 committed:
Improvied editor detection

* Fixed crash if editor cannot be found

* Fix "cannot find editor" if mimeopen is not installed

* added notepad for windows users

* Additionally:
       * check for no editor only if EDITOR is not defined
       * remove useless use of '' as default

Author:    Joshua Nelson <jyn514@gmail.com>

jyn514 committed:
Fixed crash if editor cannot be found

jyn514 committed:
Fix "cannot find editor" if mimeopen is not installed

jyn514 committed:
added notepad for windows users

jyn514 committed:
Fix codestyle

Additionally:
       * check for no editor only if EDITOR is not defined
       * remove useless use of '' as default

jyn514 committed:
Better exit codes under error conditions

Closes https://github.com/charlesdaniels/bitshuffle/issues/14

jyn514 committed:
Return error codes

Closes https://github.com/charlesdaniels/bitshuffle/issues/14

jyn514 committed:
Fix codestyle

Looking into git commit hooks so this will not happen in the future.
See https://stackoverflow.com/questions/3462955/
and https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks

jyn514 committed:
Refactored do_test to be much easier to add to

Added function do_test which abstracts most of the
necessary boilerplate

jyn514 committed:
Added double encode/decode test

jyn514 committed:
Added tests for python 2 and 3

jyn514 committed:
Add gzip compression

Add argument for compression type
Add support for python2 gzip
Add parameters to encoding
Add description of compression level
Refactor get_editor() into its own function

charlesdaniels committed:
Integrated Aiden's PR with the current dev branch.

charlesdaniels committed:
Squelched output on smoketests which passed.

When the do_test function was generalized out, the log file was missed
and never got populated. Now the output of the program under test is
only displayed on a test failure, which makes the success output much
cleaner.

charlesdaniels committed:
Removed redundand copy of version number

jyn514 committed:
Add travis script for pyinstaller

See https://github.com/charlesdaniels/bitshuffle/issues/11
for details.

jyn514 committed:
Actually run pyinstaller

jyn514 committed:
Rename pypi script for clarity

jyn514 committed:
Change `exit` to  `sys.exit`

For some reason, exit crashes when running in wine.

jyn514 committed:
Add travis scripts

Add archive files after running pyinstaller
Cache ~/.wine directory:
        Takes forever to regenerate and doesn't change substantially
        As a side effect, python will only need to be installed on first run
Add todo message for automatic PyPI releases

jyn514 committed:
Add lint for travis config

jyn514 committed:
Fix broken travis config

jyn514 committed:
Infer encode/decode (major changes)

Assign defaults *after* inferring argument
        If arguments are specified, easier to infer
        If not, it's clear that no change was made
                (as opposed to the user specifying the default)
        Use dictionary (called 'defaults') to specify
        Note: filename has to be evaluated second because it depends
              on input

Quit if no arguments given
Many layers of checks to infer arguments
Decisions made as follows: (assuming encode/decode not passed)
        editor given        -> decode
        compresstype given  -> encode
        compresslevel given -> encode
        chunklevel given    -> encode
        filename given      -> encode
        no output:          -> encode       (why?)
        no input and interactive  -> decode (why?)
        input and not interactive -> quit with error
        input and output          -> quit with error

Other refactoring:
find_editor() is its own function (simplifies logic by a great deal)

jyn514 committed:
Check if file exists before opening

That is, the file specified by --filename

jyn514 committed:
Refactor smart infer (more)

Ease of reading:
Move smart infer to own function
Move `set defaults` to own function
 - Move said defaults to said function

Control flow:
Use one big if statement for `encode` instead of 4
Exit in one controlled place if cannot infer
Simplify some more logic

TODO:
Add tests. This really needs tests.

Misc:
Remove stderr message when choosing editor

jyn514 committed:
Fix egregious errors

infer_mode returns args
editor changed to args.editor

jyn514 committed:
Print usage by default instead of help

Difference of about 10 lines of terminal.
Prints that --help is a flag, so I doubt it'll
be terribly confusing.

charlesdaniels committed:
re-structured project directory

* BitShuffle itself is now in it's own directory so it can be a real
  Python module. This will hopefully help #27.

* Deleted 'smoketest' and created a generic 'scripts' directory.

* "smoketest/do_test.sh" has been renamed 'test_end2end.sh' for clarity.

* This will probably break 'setup.py' worse than it was already broken.

charlesdaniels committed:
Fixed incorrect path in .travis.yml for end2end

charlesdaniels committed:
Resolved #27 - setup.py now works correctly

charlesdaniels committed:
Added test_setup_py.sh script for #29

charlesdaniels committed:
test_setup_py.sh now uses sudo

cross-reference: #29

charlesdaniels committed:
set sudo: required in travis.yml

charlesdaniels committed:
added further detail on sudo:required in travis.yml

charlesdaniels committed:
Fixed test_setup_py

charlesdaniels committed:
Fixed test_setup_py (again)

It seems to be having a hard time getting to the project root.

charlesdaniels committed:
test_setup_py now uses abs project tld

charlesdaniels committed:
test_setup_py now uses readlink

Ubuntu 16 does not ship realpath. Note that this may break on UNIXes that don’t ship a compatible readlink. However this script only needs to run on the CI server, so it shouldn’t be a problem.

charlesdaniels committed:
Merge branch 'dev' of github.com:charlesdaniels/bitshuffle into dev

jyn514 committed:
Fix conflicting editor options

jyn514 committed:
Check if editor exists

Use existing find_editor function
If nonexistent, print("Editor {editor} not found") and exit with error

jyn514 committed:
Add argument tests

jyn514 committed:
Merge branch 'staging' into dev

Major changes:
	Add scripts/test_argument_options
	Quit with error if editor not found
	Rewrote argument inferral from scratch
		* Function for inferral
		* Make function for defaults
		 (so we know when they have not been specified)
		* Logic:
			- compresstype or compresslevel
			  or chunksize or filename -> encode
			- editor -> decode
			- input and output -> decode
			- input and no output -> encode
			- interactive and no input -> decode
			- no input or output -> encode
			- all other cases -> quit with error

Minor changes:
	Exit 1 if given no arguments
	Exit 4 if input not found
	Bugfixes
	find_editor() no longer takes 'args' as parameter
	when testing setup.py, only use sudo if necessary
	fail smoketest if travis lint fails
	setup.py adheres to pycodestyle

jyn514 committed:
install travis on build machine

charlesdaniels committed:
Test script portability & usability

* Michael Kropat's pure sh implementation of realpath is now used for
  path normalization. This should make it possible to use any POSIX
  system to run any of the test scripts (or at least remove their
  dependency on readlink -f/-e anyway). In particular, this lets the
  test scripts run on macOS correctly. This resolves #33.

* pycodestyle is now run by a script

* Added pre_commit_check.sh script which runs all checks/scripts that
  should be done before committing. This is as a convenience to
  developers working on BitShuffle.

charlesdaniels committed:
Fixed incorrect TLD detection in test_setup_py

charlesdaniels committed:
Added test cases for encode/decode inference

jyn514 committed:
Make test_pycodestyle less ugly

jyn514 committed:
Merge branch 'dev' of https://github.com/charlesdaniels/bitshuffle into dev

charlesdaniels committed:
--output and stdin not a tty now implies decode

charlesdaniels committed:
Merge branch 'dev' of github.com:charlesdaniels/bitshuffle into dev

jyn514 commented:
Note: various parts of the static build autorelease scripts are also untested
 - scripts/travis.sh (used for static builds)
 - `twine upload dist/*` (called by scripts/pypi.sh)
 - .travis.yml: any part of the release config

jyn514 committed:
Remove unused import

charlesdaniels committed:
Updated README for clarity/detail

* Added contribution section
* Added installation instructions
* Added example use cases
* Misc. cleanup & organization

jyn514 committed:
fix typos

charlesdaniels committed:
Added some more entries to the FAQ

charlesdaniels committed:
Fixed incorrect list formatting in README

charlesdaniels committed:
Fixed incorrect list formatting again

jyn514 committed:
make realpath easier to read

jyn514 committed:
make travis check its own file

jyn514 commented on scripts/test_argument_options.sh:
```diff
@@ -0,0 +1,92 @@
+        exit 9999
+fi
+
+# Note: of necessity, the following can only test non-interactive
```
Want to make this a new issue

jyn514 replied:
Done. See https://github.com/charlesdaniels/bitshuffle/issues/37

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -133,103 +133,104 @@ def encode_file(fhandle, chunksize, compresslevel, compresstype, filename):
         # set to True for infile to be deleted after decoding
         is_tmp = False
-        if stdin.isatty() and args.input is '/dev/stdin':
+        if stdin.isatty() and args.input == '/dev/stdin':
```
Want this in as soon as possible, even if whole review takes longer

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -133,103 +133,104 @@ def encode_file(fhandle, chunksize, compresslevel, compresstype, filename):
 
+    # Set default values
+    args = set_defaults(args)
+
```
Very happy we made these functions

jyn514 commented on .travis.yml:
```diff
@@ -1,18 +1,42 @@
+# - provider: pypi
+#     on:
+#       tags: true
+# run `travis encrypt --add deploy.password` to avoid making password public
```
Lots of untested things in this file. Maybe we could only commit things known to work.

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -133,103 +133,104 @@ def encode_file(fhandle, chunksize, compresslevel, compresstype, filename):
-                    sys.exit(1)
-
-            stderr.write("editor is %s\n" % editor)
+            stderr.write("editor is %s\n" % args.editor)
```
Do we still want this included?

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -246,6 +247,24 @@ def main():
+        return os.environ['VISUAL']
+    elif 'EDITOR' in os.environ:
+        return os.environ['EDITOR']
+    else:
```
See https://github.com/charlesdaniels/bitshuffle/issues/32. No immediate changes needed.

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -300,6 +319,62 @@ def decode(message):
+
+
+def set_defaults(args):
+
```
My only concern for this is updating the defaults is no longer automatic. If we change input and output as per https://github.com/charlesdaniels/bitshuffle/issues/28, we'll have to update the help manually. Same for any other changes. Not sure how to fix this.

jyn514 reviewed (commented):
Want to remove untested features before commiting. Lots of good things, but also lots of things that have never been run, especially [travis file](https://github.com/charlesdaniels/bitshuffle/pull/34/files/a4ced202750988db4bc9e500697a9323eb20c6ce#diff-354f30a63fb0907d4ad57269548329e3)

jyn514 committed:
make travis lint executable

jyn514 committed:
Don't install pip (since already installed)

jyn514 committed:
Fix permissions for apt/sources.list

jyn514 committed:
install for ubuntu, not debian

jyn514 committed:
Remove python version from travis config

If version is different from default on docker,
build takes nearly twice as long.
Not specifying a version speeds builds and shouldn't
matter for the program itself.

jyn514 committed:
Remove pip version from travis config

If python is language agnostic, pip needs to be as well.
Otherwise, may not be installed.

charlesdaniels committed:
travis.yml: install deps for Python 2 also

charlesdaniels committed:
travis.sh: record build dir structure after run

charlesdaniels committed:
TravisCI now specifies Python version explicitly

charlesdaniels committed:
Testing different Python versions is done right

Previously, we were using the `python2` and `python3` commands. We
should just be using `python` and letting TravisCI handle the rest. This
does mean that committers will only be testing against their system
`python`, however this is unlikely to to cause trouble, as
version-specific regressions will probably be infrequent.

charlesdaniels committed:
test_setup_py now uses system `python`

charlesdaniels committed:
Fixed incorrect bitshuffle path for test_end2end

charlesdaniels committed:
Fixed WINE usage in travis.sh

charlesdaniels committed:
travish.sh now specifies WINEPREFIX as an abspath

charlesdaniels committed:
Basic AppVeyor config and smoke tests

* This resolves #35 - the smoke tests are implemented, although the one
to tests piping input/output won't pass until #28 is resolved.

jyn514 committed:
Merge branch 'dev' of https://github.com/charlesdaniels/bitshuffle into dev

jyn514 committed:
use absolute, not relative paths for wine

jyn514 committed:
add description of "data" parameters

jyn514 committed:
don't rely on same shell being used for commands

jyn514 committed:
check wine python exists

jyn514 committed:
don't cache wine directory

jyn514 committed:
debug travis script

jyn514 committed:
change "~" to $HOME

jyn514 committed:
fix pyinstaller to uppercase

charlesdaniels committed:
Patched BitShuffle to handle stdin/stdout on win32

* set_defaults now ensures that args.input and args.output are file
  handles, now nothing in BitShuffle needs to care about paths, just
  file handles.

* This resolves #28

charlesdaniels committed:
Fixed forward-incompatible use of 'if x is file'

charlesdaniels committed:
Added AppVeyor status badge to README.rst

charlesdaniels committed:
encode_file now explicitly casts data to bytes

Note that even though all of the file handles we open() are explicitly
done in binary mode, stdin / stdout are utf8 text in Python 3.X. This
commit adds code to use the Python 3.x method (stdin.buffer.read...),
but fails over to the Python 2.X method silently if it dosen't work.

jyn514 committed:
make travis less verbose

jyn514 committed:
return filename, not basename for editor

jyn514 committed:
Swallow fewer exceptions

Changed several `except Exception`s to `except IOError`
Check `isinstance(file_type)` instead of `is not str`

jyn514 committed:
Don't assume file is text-based

jyn514 committed:
remove debugging messages

charlesdaniels committed:
Removed redundant output from test_argument_options

charlesdaniels committed:
removed example from README until format is stable

jyn514 committed:
Merge pull request #38 from charlesdaniels/appveyor

Add basic AppVeyor tests for Windows, portable stdin/stdout

charlesdaniels committed:
Merge branch 'dev' into verbiage

charlesdaniels committed:
Improved help/usage messages in bitshuffle.py.

charlesdaniels committed:
Added further detail to README

charlesdaniels committed:
Documented versioning conventions

jyn514 committed:
Add pre-commit git hook

Usage: `ln scripts/pre-commit .git/hooks`

jyn514 committed:
Add doc that only one version of python is tested

In scripts/pre_commit_check, add to .SHELLDOC
that only `which python` is used for automated tests.

jyn514 committed:
fix typos and format errors

jyn514 committed:
Add various links and details

Any issue now has link to github
Change 'Python 2 or 3' to Python ( >= 2.7 )
Add link to @charlesdaniels
Add note of issue #10
Make data specification (slightly) easier to read

charlesdaniels committed:
Added shellcheck, misc. CI improvements

* revert counting mechanism for test_pycodestyle

	* The method which was added by @jyn514 caused the loop to exit
	after the first iteration. This patch revers to the old, ugly method of
	using a temporary file to store the defect counter. We can switch to a
	different method if a better one is found, but it has to work across all
	the files in the project, not just the first one.

* add test_shellcheck.sh

* brought all existing scripts into compliance with shellcheck

* disable python 'nightly' version for travis

	* Turning off Python's nightly build as a version to test with
	because I don't want regressions in Python to cause build
	failures for us.

jyn514 committed:
Merge origin into verbiage

Fix typos
Add `pycodestyle` to dependencies
Change 'python 2 or 3' to 'python ( >= 2.7 )'
Add various links
See ffd1e503071c465f07d1f00b8fb214b1795ae42d for more details.

charlesdaniels committed:
fixed typos

charlesdaniels committed:
Merge branch 'verbiage' of github.com:charlesdaniels/bitshuffle into verbiage

jyn514 committed:
Merge pull request #40 from charlesdaniels/verbiage

Improved documentation throughout the project

jyn514 committed:
Make pypi.sh posix-compliant

remove unused variable from realpath

jyn514 committed:
don't execute input

jyn514 committed:
Grab shellcheck from debian repos

Previously, compiling from source.

jyn514 committed:
test for POSIX compatability

jyn514 committed:
make pre-commit conform to styleguide

jyn514 committed:
Merge branch 'dev' into shellcheck

jyn514 committed:
Merge pull request #41 from charlesdaniels/shellcheck

Added shellcheck, misc. CI improvements

jyn514 commented:
Waiting on https://github.com/charlesdaniels/bitshuffle/issues/11 to merge.

jyn514 committed:
Remove wine builds

jyn514 committed:
install pyinstaller

jyn514 committed:
add workaround for permissions error

jyn514 committed:
travis bugfixes

jyn514 committed:
more travis permissions workarounds

jyn514 committed:
add appveyor automated release

jyn514 committed:
fix appveyor file

jyn514 committed:
Merge branch 'appveyor' into static

If 'appveyor' doesn't get bugfixes, every build will be marked
as failed, even if appveyor build passes.

Additionally, will be nice to have all static build development in one branch.

jyn514 committed:
trying things to see if they work

jyn514 committed:
make token just for appveyor

jyn514 committed:
only upload to pypi from master branch

jyn514 committed:
Check for ~/.selected_editor

Check after $VISUAL and $EDITOR but before trying random editors
Add kate, gedit, and kwrite to list of editors

Closes [32](https://github.com/charlesdaniels/bitshuffle/issues/32).

jyn514 committed:
make find_editor windows-compatible

jyn514 committed:
encrypt token properly

charlesdaniels committed:
Merge pull request #42 from charlesdaniels/static

Add automatic builds

charlesdaniels commented:
#42 has been merged, which should resolve most of #11, if not all of it. Pending review, is there a reason to continue holding up this PR?

charlesdaniels reviewed (commented):
Looks good to me. @jyn514 needs to approve this though, as PR authors cannot approve their own PRs.

jyn514 commented on README.rst:
```diff
@@ -7,47 +7,160 @@ BitShuffle README
+
+This is not possible yet, but in the future, there will be static builds of
+BitShuffle that can be run standalone.
+See also `#11 <https://github.com/charlesdaniels/bitshuffle/issues/11>`_.
```
This has been resolved, we should add a note of it as well as a link

charlesdaniels replied:
Fixed in 
917686c

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -0,0 +1,445 @@
+    args = parser.parse_args()
+
+    # Checks if no parameters were passed
+    if len(sys.argv[1:]) == 0:
```
`if not sys.argv[1:]`

charlesdaniels replied:
fixed in 917686c

charlesdaniels replied:
Actually, I could have left it with `len()` in retrospect, as it would do the same thing.

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -0,0 +1,445 @@
+
+            args.output.flush()
+
+    elif args.decode:
```
In future, would like this to be less monolithic. Fine for now.

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -0,0 +1,445 @@
+        try:
+            # python 3
+            args.output.buffer.write(payload)
+        except Exception as e:
```
Don't want to swallow all exceptions: `except AttributeError as e:`

charlesdaniels replied:
fixed in 
917686c

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -0,0 +1,445 @@
+            sys.exit(8)
+
+    args.input.close()
+    args.output.close()
```
These are never reached.

charlesdaniels replied:
they are if `args.encode`

jyn514 replied:
duplicated this for `args.decode` in interests of merging soon. Would like to refactor in the future.

charlesdaniels committed:
Review response for PR #34

jyn514 committed:
Add and fix links in README.rst

charlesdaniels committed:
fixed a typo

charlesdaniels committed:
Merge branch 'dev' of github.com:charlesdaniels/bitshuffle into dev

charlesdaniels commented:
> Linux and Windows

Fixed in  8b7b3cd

jyn514 commented on bitshuffle/bitshuffle.py:
```diff
@@ -190,7 +190,7 @@ def main():
 
     # Checks if no parameters were passed
-    if len(sys.argv[1:]) == 0:
+    if sys.argv[1:] == 0:
```
This does not do what is intended, see [hanging build](https://travis-ci.org/charlesdaniels/bitshuffle/jobs/320272051)

charlesdaniels replied:
> Actually, I could have left it with len() in retrospect, as it would do the same thing.

Okay, maybe not lol.

jyn514 committed:
Fix bad logic

jyn514 committed:
use sys.exit, not quit

jyn514 committed:
close input and output files

jyn514 reviewed (approved):
Looks good. Can merge as soon as tests pass.
